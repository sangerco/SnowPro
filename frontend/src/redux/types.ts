export interface UserData {
    username: string;
    firstName: string;
    lastName: string;
    email: string;
    avatar: string;
    bio: string;
    videos: string[];
    photos: string[];
    favMountains: string[];
  }


export const FETCH_USER_DATA_REQUEST = 'FETCH_USER_DATA_REQUEST';
export const FETCH_USER_DATA_SUCCESS = 'FETCH_USER_DATA_SUCCESS';
export const FETCH_USER_DATA_FAILURE = 'FETCH_USER_DATA_FAILURE';


export interface FetchUserDataRequestAction {
    type: typeof FETCH_USER_DATA_REQUEST
}

export interface FetchUserDataSuccessAction {
    type: typeof FETCH_USER_DATA_SUCCESS;
    payload: UserData;
}

export interface FetchUserDataFailureAction {
    type: typeof FETCH_USER_DATA_FAILURE;
    payload: string;
}

export type UserActionTypes = 
    FetchUserDataRequestAction
    | FetchUserDataSuccessAction
    | FetchUserDataFailureAction;