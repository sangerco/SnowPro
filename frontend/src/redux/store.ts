import { Action, compose } from "redux";
import { ThunkAction, configureStore } from "@reduxjs/toolkit";
import { logger } from "redux-logger";
import rootReducer from "./reducers/rootReducer";

const composeEnhancers =
  (window as any).__REDUX_DEVTOOLS_EXTENSION_COMPOSE__ || compose;

export type RootState = ReturnType<typeof store.getState>;
export type AppDispatch = typeof store.dispatch;
export type AppThunk<ReturnType = void> = ThunkAction<
  ReturnType,
  RootState,
  unknown,
  Action<string>
>;

export const store = configureStore({
  reducer: rootReducer,
  middleware: (getDefaultMiddleware) =>
    getDefaultMiddleware().concat(logger, composeEnhancers),
});
