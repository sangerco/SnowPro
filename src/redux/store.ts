import { configureStore, combineReducers } from "@reduxjs/toolkit";
import userReducer from "./slices/userSlice";
import authReducer from "./slices/authSlice";
import messageReducer from "./slices/messageSlice";
import messageReplyReducer from "./slices/messageReplySlice";
import reviewReducer from "./slices/reviewSlice";
import reviewReplyReducer from "./slices/reviewReplySlice";
import tagReducer from "./slices/tagSlice";
import storage from "redux-persist/lib/storage";
import { persistReducer, persistStore } from "redux-persist";

const persistConfig = {
  key: "root",
  storage,
  blacklist: ["messages", "messageReplies", "reviews", "reviewReplies"],
};

const rootReducer = combineReducers({
  users: userReducer,
  auth: authReducer,
  messages: messageReducer,
  messageReplies: messageReplyReducer,
  reviews: reviewReducer,
  reviewReplies: reviewReplyReducer,
  tags: tagReducer,
});

const persistedReducer = persistReducer(persistConfig, rootReducer);

export const store = configureStore({
  reducer: persistedReducer,
});

export type RootState = ReturnType<typeof store.getState>;
export type AppDispatch = typeof store.dispatch;

export const persistor = persistStore(store);
