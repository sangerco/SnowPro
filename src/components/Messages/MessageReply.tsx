import React, { useEffect, useState } from "react";
import { useSelector, useDispatch } from "react-redux";
import {
  Button,
  Dimmer,
  Loader,
  Icon,
  Card,
  Modal,
  Grid,
  Segment,
  Rail,
  Header,
} from "semantic-ui-react";
import { Link, useParams } from "react-router-dom";
import { RootState, AppDispatch } from "../../redux/store";
import {
  fetchMessageReply,
  deleteMessageReply,
} from "../../redux/slices/messageReplySlice";
import { fetchSkiAreas } from "../../redux/slices/skiAreaSlice";

const MessageReply: React.FC = () => {
  const { id } = useParams();
  const messageReplies = useSelector(
    (state: RootState) => state.messageReplies
  );
  const dispatch = useDispatch<AppDispatch>();

  const [showDeleteModal, setShowDeleteModal] = useState(false);

  useEffect(() => {
    if (id) {
      dispatch(fetchMessageReply(id));
    }
  }, [dispatch, id]);

  useEffect(() => {
    dispatch(fetchSkiAreas());
  }, [dispatch]);

  const skiAreaState = useSelector((state: RootState) => state.skiAreas);
  const skiAreas = skiAreaState.skiAreas;

  const handleShowDeleteModal = () => {
    setShowDeleteModal(true);
  };

  const handleDeleteMessageReply = () => {
    if (id) {
      dispatch(deleteMessageReply(id));
    }
  };

  console.log(messageReplies);

  const formatDate = (date: Date) => {
    const newDate = new Date(date);
    const options: Intl.DateTimeFormatOptions = {
      year: "numeric",
      month: "long",
      day: "numeric",
    };
    return newDate.toLocaleDateString("en-US", options);
  };

  if (messageReplies.loading) {
    return (
      <div>
        <Dimmer active>
          <Loader>Loading...</Loader>
        </Dimmer>
      </div>
    );
  }

  if (messageReplies.error) {
    return <div>Error: {messageReplies.error}</div>;
  }

  if (Array.isArray(messageReplies.messageReply)) {
    return <div>Error! Message Reply cannot be displayed!</div>;
  } else if (messageReplies.messageReply) {
    const messageReply = messageReplies.messageReply;

    return (
      <div>
        <Grid>
          <Grid.Row>
            <Grid.Column width={4}>
              {skiAreaState.loading ? (
                <Dimmer active>
                  <Loader>Loading...</Loader>
                </Dimmer>
              ) : skiAreas && skiAreas.length > 0 ? (
                <Segment style={{ overflow: "auto", maxHeight: "75vh" }}>
                  <Card id="ski-area-card">
                    {" "}
                    {/* max height overflow auto */}
                    {skiAreas.map((sa) => (
                      <Card.Content key={sa.slug}>
                        <Link to={`/ski-areas/${sa.slug}`}>{sa.name}</Link>
                      </Card.Content>
                    ))}
                  </Card>
                </Segment>
              ) : skiAreaState.error ? (
                <Dimmer active>
                  <Header as="h1">
                    Error! Ski Area Data cannot be retrieved!{" "}
                    {skiAreaState.error}
                  </Header>
                </Dimmer>
              ) : null}
            </Grid.Column>
            <Grid.Column width={8}>
              <Segment raised>
                <Card fluid>
                  <Card.Content>
                    <Card.Header>{messageReply.subject}</Card.Header>
                    <Card.Meta>
                      {messageReply.senderFirstName}{" "}
                      {messageReply.senderLastName}
                    </Card.Meta>
                    <Card.Meta textAlign="right">
                      {formatDate(messageReply.createdAt)}
                    </Card.Meta>
                    <Card.Description>{messageReply.body}</Card.Description>
                  </Card.Content>
                  <Card.Content extra>
                    <div className="ui two buttons">
                      <Button basic color="red" onClick={handleShowDeleteModal}>
                        <Icon name="trash" />
                      </Button>
                      <Button basic color="blue">
                        <Link to={`/messages/${messageReply.messageId}/reply`}>
                          Reply
                        </Link>
                      </Button>
                    </div>
                  </Card.Content>
                  <Modal
                    open={showDeleteModal}
                    onClose={() => setShowDeleteModal(false)}>
                    <Modal.Content>
                      Are You Sure You Want To Delete This Message Reply?
                    </Modal.Content>
                    <Modal.Actions>
                      <Button negative onClick={handleDeleteMessageReply}>
                        Yes
                      </Button>
                      <Button onClick={() => setShowDeleteModal(false)}>
                        Cancel
                      </Button>
                    </Modal.Actions>
                  </Modal>
                </Card>
              </Segment>
            </Grid.Column>
            <Grid.Column width={4}></Grid.Column>
          </Grid.Row>
        </Grid>
      </div>
    );
  }

  return null;
};

export default MessageReply;
