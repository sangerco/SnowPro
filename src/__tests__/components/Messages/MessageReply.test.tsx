/* eslint-disable testing-library/no-node-access */
import React from "react";
import MessageReply from "../../../components/Messages/MessageReply";
import { screen, fireEvent, render } from "@testing-library/react";
import { Provider } from "react-redux";
import { BrowserRouter } from "react-router-dom";
import { store } from "../../../redux/store";

describe("test Message Reply", () => {
  it("should render without crashing", () => {
    render(
      <Provider store={store}>
        <BrowserRouter>
          <MessageReply />
        </BrowserRouter>
      </Provider>
    );
  });

  it("should match the snapshot", () => {
    const { asFragment } = render(
      <Provider store={store}>
        <BrowserRouter>
          <MessageReply />
        </BrowserRouter>
      </Provider>
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
