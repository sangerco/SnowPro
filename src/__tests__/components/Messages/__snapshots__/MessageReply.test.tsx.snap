// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test Message Reply should match the snapshot 1`] = `<DocumentFragment />`;
